### PR TITLE
colfetcher: fix index join with low workmem limit when using Streamer

### LINKED ISSUE #75703: roachtest: tpchvec/disk failed
roachtest.tpchvec/disk [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4246509&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4246509&tab=artifacts#/tpchvec/disk) on master @ [bb85ddc4fc0ad8761d7aba61315da22b433022e3](https://github.com/cockroachdb/cockroach/commits/bb85ddc4fc0ad8761d7aba61315da22b433022e3):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/tpchvec/disk/run_1
	tpchvec.go:480,tpchvec.go:564,tpchvec.go:587,test_runner.go:780: Non-zero exit code: 1
```
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*tpchvec/disk.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

### PR DESCRIPTION
**Pull request by yuzefovich**
Previously, it was possible for an index join to run into "budget
exceeded" errors when using the Streamer API because we used the
hard-coded value of 4MiB for input batch size without paying attention
to possibly lower values of `distsql_workmem` limit. For example, we
could have buffered 3MiB worth of spans to be enqueued into the Streamer
that only has 1MiB of budget overall. This is now fixed by reducing the
input size limit to be a quarter of `distsql_workmem` limit. We don't
need to do a similar change to the row-by-row engine because there we
hard-code the workmem limit to be at least 8MiB.

Fixes: #75703

Release note: None

### CODE REVIEW

**Review by rharding6373 — APPROVED** [review]
Nice work! :lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75741#-:-MurExXGBnDfxmhdb6Fp:bnahwna)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rytaft and @yuzefovich)

___
*[pkg/sql/colfetcher/index_join.go, line 515 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75741#-MurDy_RBoCfC5MgM3wi:-MurDy_RBoCfC5MgM3wj:b-69lzhz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e43753860749f4857c34f297fb8f298d4c09ad7/pkg/sql/colfetcher/index_join.go#L515)):*
> ```Go
> 			// If we have a low workmem limit, then we want to reduce the input
> 			// batch size limit - namely, we'll give it a quarter of workmem
> 			// limit. This will leave a half to the Streamer itself.
> ```

So the inputBatchSizeLimit comes out of the streamer's budget allocation? Maybe reword to say something like "This will leave half of the total workmem limit to the Streamer, since the inputBatchSizeLimit is part of the Streamer budgetLimit."



_Code quote:_
```Go
half to the Streamer
```

**Review by yuzefovich — COMMENTED** [review]
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75741#-:-MurHvcQ4DKfcaWu6Tu9:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rharding6373 and @rytaft)

___
*[pkg/sql/colfetcher/index_join.go, line 515 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75741#-MurDy_RBoCfC5MgM3wi:-MurHs2qEyUEBe3-0PmZ:bzs9k3) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e43753860749f4857c34f297fb8f298d4c09ad7/pkg/sql/colfetcher/index_join.go#L515)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

So the inputBatchSizeLimit comes out of the streamer's budget allocation? Maybe reword to say something like "This will leave half of the total workmem limit to the Streamer, since the inputBatchSizeLimit is part of the Streamer budgetLimit."
</blockquote></details>

Rephrased the comment to provide more context.
